plugins/woocommerce/src/StoreApi/Utilities/CartController.php

```diff
@@ -27,20 +26,40 @@ class CartController {
 	 * Makes the cart and sessions available to a route by loading them from core.
 	 */
 	public function load_cart() {
-		if ( ! did_action( 'woocommerce_load_cart_from_session' ) && function_exists( 'wc_load_cart' ) ) {
-			wc_load_cart();
+		if ( did_action( 'woocommerce_load_cart_from_session' ) ) {
+			return;
 		}
+
+		// Initialize the cart.
+		wc_load_cart();
+
+		// Load cart from session.
+		$cart = $this->get_cart_instance();
+		$cart->get_cart();
 	}
 
 	/**
-	 * Recalculates the cart totals.
+	 * Gets the latest cart instance, and ensures totals have been calculated before returning.
+	 *
+	 * @return \WC_Cart
+	 */
+	public function get_cart_for_response() {
+		return did_action( 'woocommerce_after_calculate_totals' ) ? $this->get_cart_instance() : $this->calculate_totals();
+	}
+
+	/**
+	 * Recalculates the cart totals and returns the updated cart instance.
+	 *
+	 * @since 9.2.0 Calculate shipping was removed here because it's called already by calculate_totals.
+	 *
+	 * @return \WC_Cart
 	 */
 	public function calculate_totals() {
 		$cart = $this->get_cart_instance();
 		$cart->get_cart();
 		$cart->calculate_fees();
-		$cart->calculate_shipping();
 		$cart->calculate_totals();
```

Comment:
In my testing, leaving everything as is and just removing this reduced the amount of totals calculation while keeping other thing correct.